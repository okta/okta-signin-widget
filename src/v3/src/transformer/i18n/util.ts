/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
