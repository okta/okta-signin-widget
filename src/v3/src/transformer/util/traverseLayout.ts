/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
