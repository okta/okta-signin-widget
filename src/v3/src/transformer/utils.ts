/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FieldElement,
  FormBag,
  UISchemaElement,
  UISchemaLayoutType,
} from '../types';

export const createForm = (): FormBag => ({
  schema: {},
  uischema: {
    type: UISchemaLayoutType.VERTICAL,
    elements: [],
  },
  data: {},
  dataSchema: {
    fieldsToValidate: [] as string[],
  } as FormBag['dataSchema'],
});

export const removeUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement[] => (
  elements.filter((element) => (
    name !== (element as FieldElement).options?.inputMeta?.name
  ))
);

export const getUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement | undefined => (
  elements.find((element) => (
    name === (element as FieldElement).options?.inputMeta?.name
  ))
);
