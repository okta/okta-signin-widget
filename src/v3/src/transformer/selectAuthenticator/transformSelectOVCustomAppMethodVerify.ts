/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input, NextStep } from '@okta/okta-auth-js';

import { AUTHENTICATOR_KEY, IDX_STEP } from '../../constants';
import {
  AuthenticatorButtonListElement,
  ButtonElement,
  ButtonType,
  DescriptionElement,
  FieldElement,
  IdxStepTransformer,
  IWidgetContext,
  LinkElement,
  TitleElement,
  UISchemaElement,
} from '../../types';
import {
  getAuthenticatorKey, hasMinAuthenticatorOptions, loc, updateTransactionWithNextStep,
} from '../../util';
import { getUIElementWithName, removeUIElementWithName } from '../utils';
import { getAppAuthenticatorMethodButtonElements, isOnlyPushWithAutoChallenge } from './utils';

export const transformSelectOVCustomAppMethodVerify: IdxStepTransformer = ({
  transaction,
  formBag,
}) => {
  const {
    availableSteps,
    nextStep: { inputs, name: stepName, relatesTo } = {} as NextStep,
  } = transaction;
  const authenticator = inputs?.find(({ name }) => name === 'authenticator') as Input;
  if (!authenticator) {
    return formBag;
  }

  const { uischema, data } = formBag;
  const isOV = getAuthenticatorKey(transaction) === AUTHENTICATOR_KEY.OV;

  if (isOnlyPushWithAutoChallenge(authenticator.value as Input[])) {
    uischema.elements = removeUIElementWithName(
      'authenticator.methodType',
      uischema.elements as UISchemaElement[],
    );
    const autoChallenge = getUIElementWithName(
      'authenticator.autoChallenge',
      uischema.elements as UISchemaElement[],
    ) as FieldElement;
    if (autoChallenge) {
      data[autoChallenge.options.inputMeta.name] = autoChallenge.options.inputMeta.value;
    }
    data['authenticator.methodType'] = 'push';

    uischema.elements.unshift({
      type: 'Title',
      options: {
        content: isOV
          ? loc('oie.okta_verify.push.title', 'login')
          : loc('oie.custom_app.push.title', 'login'),
      },
    } as TitleElement);

    const sendPushButton: ButtonElement = {
      type: 'Button',
      label: isOV
        ? loc('oie.okta_verify.sendPushButton', 'login')
        : loc('oie.custom_app.sendPushButton', 'login'),
      options: {
        type: ButtonType.SUBMIT,
        step: stepName,
      },
    };
    uischema.elements.push(sendPushButton);
    const selectVerifyStep = availableSteps?.find(
      ({ name }) => name === IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
    );
    const shouldAddLink = hasMinAuthenticatorOptions(
      transaction,
      IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
      1, // Min # of auth options for link to display
    );
    if (selectVerifyStep && shouldAddLink) {
      const { name: selectAuthStep } = selectVerifyStep;
      const listLink: LinkElement = {
        type: 'Link',
        contentType: 'footer',
        options: {
          label: loc('oie.verification.switch.authenticator', 'login'),
          step: selectAuthStep,
          onClick: (widgetContext?: IWidgetContext): unknown => {
            if (typeof widgetContext === 'undefined') {
              return;
            }
            updateTransactionWithNextStep(transaction, selectVerifyStep, widgetContext);
          },
        },
      };
      uischema.elements.push(listLink);
    }
  } else {
    const buttonElements = getAppAuthenticatorMethodButtonElements(
      authenticator,
      stepName,
      isOV ? AUTHENTICATOR_KEY.OV : AUTHENTICATOR_KEY.CUSTOM_APP,
      relatesTo?.value?.deviceKnown,
    );
    uischema.elements = removeUIElementWithName(
      'authenticator.methodType',
      uischema.elements as UISchemaElement[],
    );
    const authenticatorListElement: AuthenticatorButtonListElement = {
      type: 'AuthenticatorButtonList',
      options: { buttons: buttonElements, dataSe: 'authenticator-verify-list' },
    };
    uischema.elements.push(authenticatorListElement);

    const titleElement: TitleElement = {
      type: 'Title',
      options: {
        content: loc('oie.select.authenticators.verify.title', 'login'),
      },
    };
    const descriptionElement: DescriptionElement = {
      type: 'Description',
      contentType: 'subtitle',
      options: {
        content: loc('oie.select.authenticators.verify.subtitle', 'login'),
      },
    };

    // Title -> Descr -> Element(s)
    uischema.elements.unshift(descriptionElement);
    uischema.elements.unshift(titleElement);
  }

  return formBag;
};
