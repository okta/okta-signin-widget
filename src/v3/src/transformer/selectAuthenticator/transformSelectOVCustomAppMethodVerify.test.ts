/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxAuthenticator } from '@okta/okta-auth-js';
import { AUTHENTICATOR_KEY, IDX_STEP } from 'src/constants';
import { getStubFormBag, getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import {
  AuthenticatorButtonElement,
  AuthenticatorButtonListElement,
  ButtonElement,
  ButtonType,
  DescriptionElement,
  FieldElement,
  LinkElement,
  TitleElement,
  WidgetProps,
} from 'src/types';

import { hasMinAuthenticatorOptions } from '../../util';
import { transformSelectOVCustomAppMethodVerify } from './transformSelectOVCustomAppMethodVerify';
import {
  getAppAuthenticatorMethodButtonElements,
  isOnlyPushWithAutoChallenge,
} from './utils';

const getMockMethodTypes = (): AuthenticatorButtonElement[] => {
  const authenticators: AuthenticatorButtonElement[] = [];
  authenticators.push({
    type: 'AuthenticatorButton',
    label: 'Get a push notification',
    options: {
      type: ButtonType.BUTTON,
      key: 'okta_verify',
      ctaLabel: 'Select',
      actionParams: {
        'authenticator.id': 'abcde1234',
        'authenticator.methodType': 'push',
      },
      step: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
    },
  });
  authenticators.push({
    type: 'AuthenticatorButton',
    label: 'Enter a code',
    options: {
      type: ButtonType.BUTTON,
      key: 'okta_verify',
      ctaLabel: 'Select',
      actionParams: {
        'authenticator.id': 'abcde1234',
        'authenticator.methodType': 'totp',
      },
      step: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
    },
  });
  return authenticators;
};

jest.mock('./utils', () => ({
  getAppAuthenticatorMethodButtonElements: jest.fn().mockImplementation(
    () => jest.fn().mockReturnValue(getMockMethodTypes()),
  ),
  isOnlyPushWithAutoChallenge: jest.fn().mockImplementation(
    () => jest.fn().mockReturnValue(false),
  ),
}));

jest.mock('../../util', () => ({
  // @ts-expect-error spreading required here for loc impl
  ...jest.requireActual('../../util'),
  hasMinAuthenticatorOptions: jest.fn().mockImplementation(
    () => jest.fn().mockReturnValue(false),
  ),
}));

describe('Transform Select OV Method Verify Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  const formBag = getStubFormBag();
  const widgetProps: WidgetProps = {};

  beforeEach(() => {
    formBag.uischema.elements = [
      {
        type: 'Field',
        options: { inputMeta: { name: 'authenticator.methodType' } },
      } as FieldElement,
      {
        type: 'Field',
        options: { inputMeta: { name: 'authenticator.autoChallenge', value: 'true' } },
      } as FieldElement,
    ];
    transaction.nextStep = {
      name: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
      inputs: [
        {
          name: 'authenticator',
          value: [{ name: 'autoChallenge', type: 'boolean', value: 'true' }],
        },
      ],
      relatesTo: {
        value: {
          methods: [{ type: 'push' }],
          key: AUTHENTICATOR_KEY.OV,
        } as IdxAuthenticator,
      },
    };
  });

  it('should not transform elements when transaction is missing inputs', () => {
    transaction.nextStep!.inputs = undefined;
    expect(transformSelectOVCustomAppMethodVerify({
      transaction,
      formBag,
      widgetProps,
    })).toEqual(formBag);
  });

  it('should not transform elements when transaction is missing methodType from input values', () => {
    transaction.nextStep!.inputs = [{ name: 'authenticator' }];
    expect(transformSelectOVCustomAppMethodVerify({
      transaction,
      formBag,
      widgetProps,
    })).toEqual(formBag);
  });

  it('should transform elements when transaction only contains push method type '
    + 'with autoChallenge option', () => {
    (isOnlyPushWithAutoChallenge as jest.Mock).mockReturnValue(true);
    (getAppAuthenticatorMethodButtonElements as jest.Mock).mockReturnValue([]);
    const updatedFormBag = transformSelectOVCustomAppMethodVerify({
      transaction,
      formBag,
      widgetProps,
    });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(3);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.okta_verify.push.title');
    expect((updatedFormBag.uischema.elements[1] as FieldElement).options.inputMeta.name)
      .toBe('authenticator.autoChallenge');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).label)
      .toBe('oie.okta_verify.sendPushButton');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.step)
      .toBe('select-authenticator-authenticate');
    expect(updatedFormBag.data)
      .toEqual({ 'authenticator.autoChallenge': 'true', 'authenticator.methodType': 'push' });
  });

  it('should transform elements when transaction only contains push method type '
    + 'with autoChallenge option and has verify with something else link', () => {
    transaction.availableSteps = [{
      name: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
    }];
    (hasMinAuthenticatorOptions as jest.Mock).mockReturnValue(true);
    (isOnlyPushWithAutoChallenge as jest.Mock).mockReturnValue(true);
    (getAppAuthenticatorMethodButtonElements as jest.Mock).mockReturnValue([]);
    const updatedFormBag = transformSelectOVCustomAppMethodVerify({
      transaction,
      formBag,
      widgetProps,
    });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(4);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.okta_verify.push.title');
    expect((updatedFormBag.uischema.elements[1] as FieldElement).options.inputMeta.name)
      .toBe('authenticator.autoChallenge');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).label)
      .toBe('oie.okta_verify.sendPushButton');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.step)
      .toBe('select-authenticator-authenticate');
    expect(updatedFormBag.data)
      .toEqual({ 'authenticator.autoChallenge': 'true', 'authenticator.methodType': 'push' });
    expect((updatedFormBag.uischema.elements[3] as LinkElement).type).toBe('Link');
    expect((updatedFormBag.uischema.elements[3] as LinkElement).options.label)
      .toBe('oie.verification.switch.authenticator');
  });

  it('should transform elements when transaction contains push and totp method types', () => {
    formBag.uischema.elements = [{
      type: 'Field',
      options: {
        inputMeta: {
          name: 'authenticator.methodType',
          options: [{ label: 'Enter code', value: 'totp' }, { label: 'Push', value: 'push' }],
        },
      },
    } as FieldElement];
    transaction.nextStep = {
      name: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
      inputs: [
        {
          name: 'authenticator',
          value: [{
            name: 'id',
            required: true,
            mutable: false,
            value: 'abcde1234',
          }, {
            name: 'methodType',
            options: [{ label: 'Enter code', value: 'totp' }, { label: 'Push', value: 'push' }],
          }],
        },
      ],
      relatesTo: {
        value: {
          methods: [{ type: 'push' }],
          key: AUTHENTICATOR_KEY.OV,
        } as IdxAuthenticator,
      },
    };
    (isOnlyPushWithAutoChallenge as jest.Mock).mockReturnValue(false);
    (getAppAuthenticatorMethodButtonElements as jest.Mock).mockReturnValue(getMockMethodTypes());
    const updatedFormBag = transformSelectOVCustomAppMethodVerify({
      transaction,
      formBag,
      widgetProps,
    });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(3);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.select.authenticators.verify.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).type).toBe('Description');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.select.authenticators.verify.subtitle');
    expect((updatedFormBag.uischema.elements[2] as AuthenticatorButtonElement).type).toBe('AuthenticatorButtonList');
    expect(((updatedFormBag.uischema.elements[2] as AuthenticatorButtonListElement)
      .options.dataSe)).toBe('authenticator-verify-list');
    expect(((updatedFormBag.uischema.elements[2] as AuthenticatorButtonListElement)
      .options.buttons.length)).toBe(2);
    expect(((updatedFormBag.uischema.elements[2] as AuthenticatorButtonListElement)
      .options.buttons[0] as AuthenticatorButtonElement)
      .options.actionParams?.['authenticator.methodType']).toBe('push');
    expect(((updatedFormBag.uischema.elements[2] as AuthenticatorButtonListElement)
      .options.buttons[0] as AuthenticatorButtonElement).label)
      .toBe('Get a push notification');
    expect(((updatedFormBag.uischema.elements[2] as AuthenticatorButtonListElement)
      .options.buttons[1] as AuthenticatorButtonElement)
      .options.actionParams?.['authenticator.methodType']).toBe('totp');
    expect(((updatedFormBag.uischema.elements[2] as AuthenticatorButtonListElement)
      .options.buttons[1] as AuthenticatorButtonElement).label).toBe('Enter a code');
  });
});

describe('Transform Select Custom App Method Verify Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  const formBag = getStubFormBag();
  const widgetProps: WidgetProps = {};

  beforeEach(() => {
    formBag.uischema.elements = [
      {
        type: 'Field',
        options: { inputMeta: { name: 'authenticator.methodType' } },
      } as FieldElement,
      {
        type: 'Field',
        options: { inputMeta: { name: 'authenticator.autoChallenge', value: 'true' } },
      } as FieldElement,
    ];
    transaction.nextStep = {
      name: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
      inputs: [
        {
          name: 'authenticator',
          value: [{ name: 'autoChallenge', type: 'boolean', value: 'true' }],
        },
      ],
      relatesTo: {
        value: {
          methods: [{ type: 'push' }],
          key: AUTHENTICATOR_KEY.CUSTOM_APP,
        } as IdxAuthenticator,
      },
    };
  });

  it('should transform elements when transaction only contains push method type '
    + 'with autoChallenge option and Custom App authenticator key', () => {
    (isOnlyPushWithAutoChallenge as jest.Mock).mockReturnValue(true);
    (getAppAuthenticatorMethodButtonElements as jest.Mock).mockReturnValue([]);
    const updatedFormBag = transformSelectOVCustomAppMethodVerify({
      transaction,
      formBag,
      widgetProps,
    });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(3);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.custom_app.push.title');
    expect((updatedFormBag.uischema.elements[1] as FieldElement).options.inputMeta.name)
      .toBe('authenticator.autoChallenge');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).label)
      .toBe('oie.custom_app.sendPushButton');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.step)
      .toBe('select-authenticator-authenticate');
    expect(updatedFormBag.data)
      .toEqual({ 'authenticator.autoChallenge': 'true', 'authenticator.methodType': 'push' });
  });

  it('should transform elements when transaction only contains push method type '
    + 'with Custom App authenticator key, autoChallenge option, and has verify with something else link', () => {
    transaction.availableSteps = [{
      name: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
    }];
    (hasMinAuthenticatorOptions as jest.Mock).mockReturnValue(true);
    (isOnlyPushWithAutoChallenge as jest.Mock).mockReturnValue(true);
    (getAppAuthenticatorMethodButtonElements as jest.Mock).mockReturnValue([]);
    const updatedFormBag = transformSelectOVCustomAppMethodVerify({
      transaction,
      formBag,
      widgetProps,
    });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(4);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.custom_app.push.title');
    expect((updatedFormBag.uischema.elements[1] as FieldElement).options.inputMeta.name)
      .toBe('authenticator.autoChallenge');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).label)
      .toBe('oie.custom_app.sendPushButton');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
    expect((updatedFormBag.uischema.elements[2] as ButtonElement).options?.step)
      .toBe('select-authenticator-authenticate');
    expect(updatedFormBag.data)
      .toEqual({ 'authenticator.autoChallenge': 'true', 'authenticator.methodType': 'push' });
    expect((updatedFormBag.uischema.elements[3] as LinkElement).type).toBe('Link');
    expect((updatedFormBag.uischema.elements[3] as LinkElement).options.label)
      .toBe('oie.verification.switch.authenticator');
  });
});
