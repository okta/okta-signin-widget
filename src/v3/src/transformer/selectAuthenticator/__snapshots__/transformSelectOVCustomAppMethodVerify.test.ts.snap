// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform Select Custom App Method Verify Tests should transform elements when transaction only contains push method type with Custom App authenticator key, autoChallenge option, and has verify with something else link 1`] = `
Object {
  "data": Object {
    "authenticator.autoChallenge": "true",
    "authenticator.methodType": "push",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.custom_app.push.title",
        },
        "type": "Title",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "authenticator.autoChallenge",
            "value": "true",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.custom_app.sendPushButton",
        "options": Object {
          "step": "select-authenticator-authenticate",
          "type": "submit",
        },
        "type": "Button",
      },
      Object {
        "contentType": "footer",
        "options": Object {
          "label": "oie.verification.switch.authenticator",
          "onClick": [Function],
          "step": "select-authenticator-authenticate",
        },
        "type": "Link",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Transform Select Custom App Method Verify Tests should transform elements when transaction only contains push method type with autoChallenge option and Custom App authenticator key 1`] = `
Object {
  "data": Object {
    "authenticator.autoChallenge": "true",
    "authenticator.methodType": "push",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.custom_app.push.title",
        },
        "type": "Title",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "authenticator.autoChallenge",
            "value": "true",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.custom_app.sendPushButton",
        "options": Object {
          "step": "select-authenticator-authenticate",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Transform Select OV Method Verify Tests should transform elements when transaction contains push and totp method types 1`] = `
Object {
  "data": Object {
    "authenticator.autoChallenge": "true",
    "authenticator.methodType": "push",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.select.authenticators.verify.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.select.authenticators.verify.subtitle",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "buttons": Array [
            Object {
              "label": "Get a push notification",
              "options": Object {
                "actionParams": Object {
                  "authenticator.id": "abcde1234",
                  "authenticator.methodType": "push",
                },
                "ariaLabel": "Select to get a push notification.",
                "ctaLabel": "Select",
                "key": "okta_verify",
                "step": "select-authenticator-authenticate",
                "type": "button",
              },
              "type": "AuthenticatorButton",
            },
            Object {
              "label": "Enter a code",
              "options": Object {
                "actionParams": Object {
                  "authenticator.id": "abcde1234",
                  "authenticator.methodType": "totp",
                },
                "ariaLabel": "Select to enter a code.",
                "ctaLabel": "Select",
                "key": "okta_verify",
                "step": "select-authenticator-authenticate",
                "type": "button",
              },
              "type": "AuthenticatorButton",
            },
          ],
          "dataSe": "authenticator-verify-list",
        },
        "type": "AuthenticatorButtonList",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Transform Select OV Method Verify Tests should transform elements when transaction only contains push method type with autoChallenge option 1`] = `
Object {
  "data": Object {
    "authenticator.autoChallenge": "true",
    "authenticator.methodType": "push",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.okta_verify.push.title",
        },
        "type": "Title",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "authenticator.autoChallenge",
            "value": "true",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.okta_verify.sendPushButton",
        "options": Object {
          "step": "select-authenticator-authenticate",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Transform Select OV Method Verify Tests should transform elements when transaction only contains push method type with autoChallenge option and has verify with something else link 1`] = `
Object {
  "data": Object {
    "authenticator.autoChallenge": "true",
    "authenticator.methodType": "push",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.okta_verify.push.title",
        },
        "type": "Title",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "authenticator.autoChallenge",
            "value": "true",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.okta_verify.sendPushButton",
        "options": Object {
          "step": "select-authenticator-authenticate",
          "type": "submit",
        },
        "type": "Button",
      },
      Object {
        "contentType": "footer",
        "options": Object {
          "label": "oie.verification.switch.authenticator",
          "onClick": [Function],
          "step": "select-authenticator-authenticate",
        },
        "type": "Link",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
